File: 数组/优先堆(最大堆)&贪心.c
class Solution {
public:
    double mincostToHireWorkers(vector<int>& quality, vector<int>& wage, int k) {
        int n = quality.size(), q_sum = 0;
        vector<int> idx(n);
        std::iota(idx.begin(), idx.end(), 0);  // 递增0，idx是 wage / quality 的顺序。对于 int[n] 不适用
        ranges::sort(idx, [&](int i, int j) {
            return wage[i] * quality[j] < wage[j] * quality[i];  // 用 / 会错，原因？
        });
        priority_queue<int, vector<int>, less<>> pq;  // 等同 priority_queue<int>，C++默认为最大堆
        for (int i = 0; i < k; i++) {
            pq.push(quality[idx[i]]);
            q_sum += quality[idx[i]];
        }

        double res = q_sum * ((double) wage[idx[k - 1]] / quality[idx[k - 1]]);  // 转为 double

        for (int i = k; i < n; i++) {
            int q = quality[idx[i]];
            if (pq.top() > q) {
                q_sum -= pq.top() - q;
                pq.pop();
                pq.push(q);
                res = min(res, q_sum * ((double) wage[idx[i]] / q));
            }
        }
        return res;
    }
};

File: 数组/优先堆(最大堆)&贪心.java
class Solution {
    public double mincostToHireWorkers(int[] quality, int[] wage, int k) {
        int n = quality.length;
        Integer[] idx = new Integer[n];
        Arrays.setAll(idx, i -> i);  // 设置值，i为索引值
        Arrays.sort(idx, (i, j) -> wage[i] * quality[j] - wage[j] * quality[i]);

        int q_sum = 0;
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);  // 默认为最小堆，b - a为最大堆
        for (int i = 0; i < k; i++) {
            pq.offer(quality[idx[i]]);
            q_sum += quality[idx[i]];
        }

        double res = q_sum * ((double) wage[idx[k - 1]] / quality[idx[k - 1]]);
        for (int i = k; i < n; i++) {
            int q = quality[idx[i]];
            if (pq.peek() > q) {
                q_sum -= pq.poll() - q;
                pq.offer(q);
                res = Math.min(res, q_sum * ((double) wage[idx[i]] / quality[idx[i]]));
            }
        }
        return res;
    }
}
